tools/stimulus/rinterface.c.in

```diff
@@ -2398,23 +2398,37 @@ void R_igraph_warning_handler(const char *reason, const char *file,
 }
 
 extern int R_interrupts_pending;
-extern int R_interrupts_suspended;
+
+void checkInterruptFn(void *dummy) {
+  IGRAPH_UNUSED(dummy);
+  R_CheckUserInterrupt();
+}
 
 int R_igraph_interrupt_handler(void *data) {
   /* We need to call R_CheckUserInterrupt() regularly to enable interruptions.
    * However, if an interruption is pending, R_CheckUserInterrupt() will
    * longjmp back to the top level so we cannot clean up ourselves by calling
-   * IGRAPH_FINALLY_FREE(). Therefore, we first check whether there are any
-   * pending interruptions. If so, and interruptions are not suspended, we call
-   * IGRAPH_FINALLY_FREE(), knowing that the upcoming invocation of
-   * R_CheckUserInterrupt() will longjmp. This means that the conditions used
-   * here must be kept in sync with the source code of R_CheckUserInterrupt()
+   * IGRAPH_FINALLY_FREE(). Therefore, we call R_CheckUserInterrupt()
+   * encapsulated in checkInterruptFn(), called through R_ToplevelExec(). If
+   * an interruption is pending, the function will properly return here instead
+   * of doing a longjmp all the way to the top. If an interruption was indeed
+   * pending, we then call IGRAPH_FINALLY_FREE(), knowing that the upcoming
+   * invocation of R_CheckUserInterrupt() will longjmp. However, we need to
+   * make sure that R_interrupts_pending = 1, in order to make sure that the
+   * interrupt will longjmp. This means that the conditions used here must be
+   * kept in sync with the source code of R_CheckUserInterrupt()
    */
-  if (!R_interrupts_suspended && R_interrupts_pending) {
+  if (R_ToplevelExec(checkInterruptFn, NULL) == FALSE) {
     IGRAPH_FINALLY_FREE();
+    /* We keep looping until R_CheckUserInterrupt will longjmp */
+    while (1)
```

Comment:
Do you have any evidence that this loop may execute more than once? If not, can we go with the fatal error approach? That way we might get some feedback from users, and some indication of when (or in what R version) things go wrong.

If we become confident that the loop is necessary (unlikely), we can still change this in a future 1.3.x release.

Reply:
No, I haven't seen this behaviour in practice, but I think something like this might arise when for example the CPU is close to 100% occupied (not necessarily by R) and things start to freeze and `R_interrupts_suspended = 1` when calling `R_CheckUserInterrupt`. Alternatively, we could perhaps do this:

```suggestion
    R_interrupts_pending = 1;
    R_CheckUserInterrupt();

    /* Process was not yet interrupted, retry */
    sleep(1);
    R_interrupts_pending = 1;
    R_CheckUserInterrupt();

    IGRAPH_FATAL("Interruption failed");
```

If the first attempt at interruption fails, we try again after a second, and then just bail out if that still didn't work.

Reply:
Let's not try to fix a problem that we don't know exists. Let's find out if it exists first. The way to do that is to just insert the `IGRAPH_FATAL`, without a `sleep`. No harm will come from this. It just calls `error()` which we even considered as a solution. But it does print a warning message to the user, and it does tell them to go and file a bug, which hopefully they will. Then we'll find out if this is a real problem in practice.

Side note: In fact, I don't even think that R sets interruption from a different thread on Windows. It appears to run the GUI within the same thread. If we don't call `R_ProcessEvents()`, the GUI will completely lock up while R is busy.

Reply:
If we do see complaints about it, we can look for ways to fix it.